Support using SYCL in C++ module

C++ modules are gaining more and more support from compilers and CMake. Currently trying to use AdaptiveCpp in C++ module leads to a link error with LLVM 19 and AdaptiveCpp 24.10.0+git.3ac3c8fb.20250210.branch.develop (see example project below):

```
undefined reference to `__acpp_reflection_init_registered_function_pointers'
```

It would be nice to be able to modularize codebase that uses AdaptiveCpp.

**Example project**

main.cpp:

```cpp
import examplemodule;

int main() {
    ls_platforms();
}

```

examplemodule.cpp:

```cpp
module;

#include <print>

#include "sycl/sycl.hpp"

export module examplemodule;

export void
ls_platforms() {
    // Loop through available platforms
    for (auto const& platform : sycl::platform::get_platforms()) {
        std::println("Found platform: {}", platform.get_info<sycl::info::platform::name>());
        // Loop through available devices in this platform
        for (auto const& device : platform.get_devices()) {
            std::println("With device: ", device.get_info<sycl::info::device::name>());
        }
        std::println("");
    }
}

```

CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.30 FATAL_ERROR)

project(
    sycl-adaptivecpp-example
    VERSION 0.0.1
    DESCRIPTION ""
    HOMEPAGE_URL "test@foo.com"
    LANGUAGES CXX
)

find_package(AdaptiveCpp REQUIRED)

add_library(examplemodule STATIC)
target_sources(examplemodule
    PUBLIC FILE_SET CXX_MODULES FILES 
    examplemodule.cpp
)

add_sycl_to_target(TARGET examplemodule)

# Tell CMake we're using C++26 and that C++26 is needed to consume it.
target_compile_features(examplemodule
    PRIVATE
    cxx_std_26
    INTERFACE
    cxx_std_26
)

# Make an executable.
add_executable(main)

target_sources(main
  PRIVATE
    main.cpp
)

target_link_libraries(main PRIVATE examplemodule)
```

Build with: `cmake -GNinja -Bbuild -DCMAKE_CXX_COMPILER=clang++ && ninja -C build`

> undefined reference to `__acpp_reflection_init_registered_function_pointers'

This function is supposed to be handled by the AdaptiveCpp LLVM IR passes, so it looks like AdaptiveCpp components are not being run correctly. Perhaps cmake does not actually use AdaptiveCpp for parts of the compilation process when modules are involved.

Can you try whether your example works when you compile it manually by directly invoking `acpp`?

Trying to compile files `examplemodule.cpp`:

```cpp
module;

#include <print>
#include "sycl/sycl.hpp"

export module examplemodule;

export void
ls_platforms() {
    // Loop through available platforms
    for (auto const& platform : sycl::platform::get_platforms()) {
        std::println("Found platform: {}", platform.get_info<sycl::info::platform::name>());
        // Loop through available devices in this platform
        for (auto const& device : platform.get_devices()) {
            std::println("With device: ", device.get_info<sycl::info::device::name>());
        }
        std::println("");
    }
}

```

 and `main.cpp`:

```cpp
import examplemodule;

int main() {
    ls_platforms();
}
```

with:

```shell
> acpp -c -fmodule-output -std=c++26 -x c++-module examplemodule.cpp
> acpp -std=c++26 -fprebuilt-module-path=. examplemodule.pcm main.cpp
```

First command produces `examplemodule.o` and `examplemodule.pcm` but the second will give:

```
/usr/bin/ld: /tmp/examplemodule-8213c2.o: in function `__acpp_reflection_tu_init_trigger::__acpp_reflection_tu_init_trigger()':
examplemodule.pcm:(.text._ZN33__acpp_reflection_tu_init_triggerC2Ev[_ZN33__acpp_reflection_tu_init_triggerC2Ev]+0x5): undefined reference to `__acpp_reflection_init_registered_function_pointers'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

If everything related to sycl is removed from `examplemodule.cpp`, then the compilation is successful with the same commands.

--- a/include/hipSYCL/glue/reflection.hpp
+++ b/include/hipSYCL/glue/reflection.hpp
@@ -73,13 +73,6 @@ const char* resolve_function_name(Ret (*func)(Args...)) {
 extern "C" void __acpp_reflection_init_registered_function_pointers();
 extern "C" void __acpp_function_annotation_needs_function_ptr_argument_reflection();
 
-// Compiler will generate calls to this function to register functions.
-__attribute__((used))
-inline void __acpp_reflection_associate_function_pointer(const void *func_ptr,
-                                             const char *func_name) {
-  hipsycl::rt::support::symbol_information::get()
-      .register_function_symbol(func_ptr, func_name);
-}
 
 struct __acpp_reflection_tu_init_trigger {
   __acpp_reflection_tu_init_trigger() {

--- a/include/hipSYCL/runtime/support/reflection.hpp
+++ b/include/hipSYCL/runtime/support/reflection.hpp
@@ -10,7 +10,6 @@
 // SPDX-License-Identifier: BSD-2-Clause
 #ifndef HIPSYCL_SUPPORT_REFLECTION_HPP
 #define HIPSYCL_SUPPORT_REFLECTION_HPP
-
 #include <mutex>
 #include <unordered_map>
 

--- a/src/compiler/reflection/FunctionNameExtractionPass.cpp
+++ b/src/compiler/reflection/FunctionNameExtractionPass.cpp
@@ -23,6 +23,8 @@
 #include <llvm/IR/GlobalVariable.h>
 #include <llvm/Support/Alignment.h>
 
+
+
 namespace hipsycl {
 namespace compiler {
 
@@ -61,6 +63,21 @@ bool isAnyUserReflectionAnnotatedFunction(llvm::Function* F, const utils::Proces
 
 }
 
+llvm::Type* getCharPtrType(llvm::Module* M, unsigned AS = 0) {
+  #if LLVM_VERSION_MAJOR >= 16
+      return llvm::PointerType::get(M->getContext(), AS);
+  #else
+      return llvm::PointerType::get(llvm::Type::getInt8Ty(M->getContext()), AS);
+  #endif
+  }
+  llvm::Type* getVoidPtrType(llvm::Module* M, unsigned AS = 0) {
+    #if LLVM_VERSION_MAJOR >= 16
+        return llvm::PointerType::get(M->getContext(), AS);
+    #else
+        return llvm::PointerType::get(llvm::Type::getInt8Ty(M->getContext()), AS);
+    #endif
+  }
+
 llvm::PreservedAnalyses FunctionNameExtractionPass::run(llvm::Module &M, llvm::ModuleAnalysisManager &AM) {
   std::string GVPrefix = "__acpp_functioninfo_";
 
@@ -82,11 +99,27 @@ llvm::PreservedAnalyses FunctionNameExtractionPass::run(llvm::Module &M, llvm::M
     }
   }
 
+  //declare explicitly __acpp_reflection_associate_function_pointer
+  static const char* ReflectionAssociateFunctionPointer = "__acpp_reflection_associate_function_pointer";
+  llvm::SmallVector<llvm::Type*> ParamTs;
+  // function pointer
+  ParamTs.push_back(getVoidPtrType(&M));
+  // function name
+  ParamTs.push_back(getCharPtrType(&M));  
+  // declare if not already declared
+  auto FC = M.getOrInsertFunction(ReflectionAssociateFunctionPointer,
+                                  llvm::FunctionType::get(llvm::Type::getVoidTy(M.getContext()),
+                                                          llvm::ArrayRef<llvm::Type *>{ParamTs},
+                                                          false));
+  llvm::Function *NewDeclaration = llvm::dyn_cast<llvm::Function>(FC.getCallee());
+  std::string NewDeclarationName = NewDeclaration->getName().str();
+  
+  
   llvm::Function* MapFunc = nullptr;
   for(auto& F : M)
     if(F.getName().contains("__acpp_reflection_associate_function_pointer"))
       MapFunc = &F;
-
+  
   if(MapFunc) {
     if(auto* InitFunc = M.getFunction("__acpp_reflection_init_registered_function_pointers")){
       if(InitFunc->isDeclaration() && (MapFunc->getFunctionType()->getNumParams() == 2)) {

--- a/src/runtime/support/reflection.cpp
+++ b/src/runtime/support/reflection.cpp
@@ -18,3 +18,10 @@ namespace hipsycl::rt::support {
     return si;
   }
 }
+
+// Compiler will generate calls to this function to register functions
+void __acpp_reflection_associate_function_pointer(const void *func_ptr,
+                                           const char *func_name) {
+  hipsycl::rt::support::symbol_information::get()
+      .register_function_symbol(func_ptr, func_name);
+}

--- a/src/runtime/support/reflection.cpp
+++ b/src/runtime/support/reflection.cpp
@@ -19,10 +19,9 @@ namespace hipsycl::rt::support {
   }
 }
 
-//avoid cpp name mangling
-extern "C" void __acpp_reflection_associate_function_pointer(const void*, const char*);
+
 // Compiler will generate calls to this function to register functions
-void __acpp_reflection_associate_function_pointer(const void *func_ptr,
+extern "C" void __acpp_reflection_associate_function_pointer(const void *func_ptr,
                                            const char *func_name) {
   hipsycl::rt::support::symbol_information::get()
       .register_function_symbol(func_ptr, func_name);

--- a/src/runtime/support/reflection.cpp
+++ b/src/runtime/support/reflection.cpp
@@ -19,6 +19,8 @@ namespace hipsycl::rt::support {
   }
 }
 
+//avoid cpp name mangling
+extern "C" void __acpp_reflection_associate_function_pointer(const void*, const char*);
 // Compiler will generate calls to this function to register functions
 void __acpp_reflection_associate_function_pointer(const void *func_ptr,
                                            const char *func_name) {

